server/app/controllers/admin/AdminApplicationController.java

```diff
@@ -148,16 +154,40 @@ public Result downloadDemographics() {
         .withHeader("Content-Disposition", String.format("attachment; filename=\"%s\"", filename));
   }
 
-  /** Download a PDF file of the application to the program. This feature is not implemented yet. */
+  /** Download a PDF file of the application to the program. */
   @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
   public Result download(Http.Request request, long programId, long applicationId) {
     try {
       ProgramDefinition program = programService.getProgramDefinition(programId);
       checkProgramAdminAuthorization(profileUtils, request, program.adminName()).join();
-      throw new UnsupportedOperationException("Not yet implemented.");
     } catch (ProgramNotFoundException e) {
       return notFound(e.toString());
+    } catch (CompletionException e) {
+      return unauthorized();
+    }
+
+    Optional<Application> applicationMaybe =
+        this.applicationRepository.getApplication(applicationId).toCompletableFuture().join();
+
+    if (!applicationMaybe.isPresent()) {
+      return notFound(String.format("Application %d does not exist.", applicationId));
+    }
+
+    Application application = applicationMaybe.get();
+    String applicantNameWithApplicationId =
+        String.format("%s (%d)", application.getApplicantData().getApplicantName(), application.id);
+
+    String filename =
+        String.format("%s-%s.pdf", applicantNameWithApplicationId, clock.instant().toString());
+    byte[] pdf;
```

Comment:
suggestion (blocking): formatting the applicant's name and the PDF filename are not HTTP concerns so probably would be best encapsulated somewhere other than the controller. Since they're fairly specific to PDF export, the `PdfExporter` may be a good place for them.

A way to do cleaner separation of concerns here could be to have `PdfExporter.export` return an instance of a new class `InMemoryPdf` that has getters for `byte [] getBytes()`, `String getFilename()`. A convenient place to define the new class could be as an inner class of the exporter:
```java
class PdfExporter {
...
  public static class InMemoryPdf { ... }
```

This would keep the details of preparing the PDF file in the exporter and keep the controller focused on HTTP concerns.